File: localsearch.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h> 
#include <string.h>
#include <math.h>
#include "karp.c"

void newInput(unsigned long long* newinput, int size, int* solution, unsigned long long* array_A) {
  int j; 

  // clear new input
  for (j = 0; j<size; j++){
      newinput[j] = 0; 
  }

  for (j = 0; j<size; j++)
    newinput[solution[j]] += array_A[j]; 
}

float t_iteration(int i) {
  return pow(10, 10) * pow(.8, (float)i/300);
}

int* copyRandomSolution(int* original_random_solution, int size) {
  int* copied_random_solution = malloc(sizeof(int) * size);

  int i;
  for (i = 0; i < size; i++) {
    copied_random_solution[i] = original_random_solution[i];
  }
  return copied_random_solution;
}

int* hillClimbingLocalSearch(unsigned long long* array_A, int* random_solution, int size, int max_iteration) {

	// logic to create A' list
	unsigned long long* newinput = calloc(1, sizeof(unsigned long long) * size); 
	struct minHeap* heap = createMinHeap(size);
  // get prior residue
  newInput(newinput, size, random_solution, array_A); 
  initializeMinHeap(heap, newinput, size); 
  unsigned long long prior_residue = karp(heap, size);


	int i;
	for (i = 0; i < max_iteration; i++) {
		int* swapped_elements = randomMove(random_solution, size);
		newInput(newinput, size, random_solution, array_A); 
    
		initializeMinHeap(heap, newinput, size); 
		unsigned long long new_residue = karp(heap, size);
		if (new_residue < prior_residue) {
			prior_residue = new_residue;
      printf("test2 %llu\n", prior_residue);
    }
    else
      swapArray(random_solution, swapped_elements[0], swapped_elements[1]);

    free(swapped_elements);
	}

	freeHeap(heap);
	free(newinput);

	return random_solution;
}

int* repeatedRandomLocalSearch(unsigned long long* array_A, int* random_solution, int size, int max_iteration) {

  // logic to create A' list
  unsigned long long* newinput = calloc(1, sizeof(unsigned long long) * size); 
  struct minHeap* heap = createMinHeap(size);
  // get prior residue
  newInput(newinput, size, random_solution, array_A); 
  initializeMinHeap(heap, newinput, size); 
  unsigned long long prior_residue = karp(heap, size);  

  int i;
  for (i = 0; i < max_iteration; i++) {
    int* new_random_solution = randomSol(size);
    newInput(newinput, size, new_random_solution, array_A); 
    
    initializeMinHeap(heap, newinput, size); 
    unsigned long long new_residue = karp(heap, size);
    if (new_residue < prior_residue) {
      int* temp = random_solution;
      random_solution = new_random_solution;
      free(temp);
      prior_residue = new_residue;
      printf("test1 %llu\n", prior_residue);
    } 
    else {
      free(new_random_solution);
    }
      
  }

  freeHeap(heap);
  free(newinput);

  return random_solution;
}

int* simulatedAnnealingLocalSearch(unsigned long long* array_A, int* random_solution, int size, int max_iteration) {

  // logic to create A' list
  unsigned long long* newinput = calloc(1, sizeof(unsigned long long) * size); 
  struct minHeap* heap = createMinHeap(size);
  // get prior residue
  newInput(newinput, size, random_solution, array_A); 
  initializeMinHeap(heap, newinput, size); 
  unsigned long long prior_residue = karp(heap, size);

  // S'' residue
  unsigned long long original_residue = prior_residue;
  // random_solution2 = S''
  int* random_solution2 = copyRandomSolution(random_solution, size);

  int i;
  for (i = 0; i < max_iteration; i++) {
    int* swapped_elements = randomMove(random_solution, size);
    newInput(newinput, size, random_solution, array_A); 

    initializeMinHeap(heap, newinput, size); 
    unsigned long long new_residue = karp(heap, size);
    if (new_residue < prior_residue) {
      prior_residue = new_residue;
      printf("test3 moved to better neighbor %llu\n", prior_residue);

    }
    else if ((rand() / (double)RAND_MAX) < exp(-(new_residue - prior_residue)/t_iteration(i))) {
      prior_residue = new_residue;
      printf("test3 random move %llu\n", prior_residue);

    } 
    else {
      swapArray(random_solution, swapped_elements[0], swapped_elements[1]);
    }

    if (prior_residue < original_residue) {
      free(random_solution2);
      random_solution2 = copyRandomSolution(random_solution, size);
      original_residue = prior_residue;
    } 

    free(swapped_elements);
  }

  free(random_solution);
  freeHeap(heap);
  free(newinput);
  return random_solution2;
}


int main() {
	srand(time(NULL));
  int size = 100;
  writeFile(size, "input2.txt");
  unsigned long long* input = readFile("input2.txt", size); 
  struct minHeap* heap = createMinHeap(size);
  unsigned long long* newinput = calloc(1, sizeof(unsigned long long) * size); 
  int* random_solution = randomSol(size);

  // repeated random test 1
  int* solution1 = repeatedRandomLocalSearch(input, copyRandomSolution(random_solution, size), size, 10);
  newInput(newinput, size, solution1, input);
  initializeMinHeap(heap, newinput, size); 

  unsigned long long final_residue = karp(heap, size);
  unsigned long long final_residue2 = karp2(newinput, size);

  printf("Repeated Random Search: \n");
  printf("Solution1: \n"); 
  int j;
  for (j = 0; j<size; j++) {
    printf("%d, ", solution1[j]);
  }
  printf("\n");
  
  printf("The final residue from repeated random sampling was %llu\n", final_residue);
  printf("%llu\n", final_residue2);
  printf("******************************************************************\n");

  // hill climbing test 2
  int* solution2 = hillClimbingLocalSearch(input, copyRandomSolution(random_solution, size), size, 10);

  newInput(newinput, size, solution2, input);
  initializeMinHeap(heap, newinput, size); 

  final_residue = karp(heap, size);
  final_residue2 = karp2(newinput, size);

  printf("Hill Climbing Search: \n");
  printf("Solution2: \n"); 
  for (j = 0; j<size; j++) {
    printf("%d, ", solution2[j]);
  }
  printf("\n");
  
  printf("The final residue from hill climbing was %llu\n", final_residue);
  printf("%llu\n", final_residue2);
  printf("******************************************************************\n");

  // Simulated Annealing test 3
  int* solution3 = simulatedAnnealingLocalSearch(input, copyRandomSolution(random_solution, size), size, 10);

  newInput(newinput, size, solution3, input);
  initializeMinHeap(heap, newinput, size); 

  final_residue = karp(heap, size);
  final_residue2 = karp2(newinput, size);
  printf("Simulated Annealing Search: \n");
  printf("Solution3: \n"); 
  for (j = 0; j<size; j++) {
    printf("%d, ", solution3[j]);
  }
  printf("\n");
  
  printf("The final residue from simulated annealing was %llu\n", final_residue);
  printf("%llu\n", final_residue2);
  printf("******************************************************************\n");

  initializeMinHeap(heap, input, size); 
  printf("Normal karp: \n"); 
  printf("The original normal residue was %llu\n", karp(heap, size));

  free(input);
  free(newinput);
  free(solution1);
  free(solution2); 
  free(solution3);
  freeHeap(heap);

}

File: python-code/localsearch.py
from random import *
from math import *
from karp import *
import copy

def newInput(original_lst, solution, size):
	new_input = []
	for j in range(size):
		new_input.append(0)

	for j in range(size):
		new_input[solution[j]] += original_lst[j]

	return new_input

def t_iteration(iter):
	return pow(10, 10) * pow(.8, float(iter)/300)


def hillClimbingLocalSearch(original_lst, random_solution, size, max_iteration):
	# random_solution = randomSol(size)
	new_input = newInput(original_lst, random_solution, size)

	prior_residue = karp(new_input, size)

	for i in range(max_iteration):
		# new_random_solution = random_solution
		new_random_solution = randomMove(random_solution, size)

		new_input2 = newInput(original_lst, new_random_solution, size)

		new_residue = karp(new_input2, size)

		if new_residue < prior_residue:
			prior_residue = new_residue
			random_solution = new_random_solution

	return random_solution


def simulatedAnnealingLocalSearch(original_lst, random_solution, size, max_iteration):
	# random_solution = randomSol(size)
	random_solution2 = random_solution

	new_input = newInput(original_lst, random_solution, size)

	prior_residue = karp(new_input, size)

	original_residue = prior_residue

	for i in range(max_iteration):
		new_random_solution = randomMove(random_solution, size)
		new_input2 = newInput(original_lst, new_random_solution, size)

		new_residue = karp(new_input2, size)

		if new_residue < prior_residue:
			prior_residue = new_residue
			random_solution = new_random_solution
		else:
			if random.random() < exp(-(new_residue - prior_residue)/float(t_iteration(i))):
				prior_residue = new_residue
				random_solution = new_random_solution

		if prior_residue < original_residue:
			random_solution2 = random_solution
			original_residue = prior_residue

	return random_solution2

def repeatedRandomLocalSearch(original_lst, random_solution, size, max_iteration):
	# random_solution = randomSol(size)
	new_input = newInput(original_lst, random_solution, size)

	prior_residue = karp(new_input, size)

	for i in range(max_iteration):
		new_random_solution = randomSol(size)
		new_input2 = newInput(original_lst, new_random_solution, size)

		new_residue = karp(new_input2, size)

		if new_residue < prior_residue:
			prior_residue = new_residue
			random_solution = new_random_solution

	return random_solution


#***********************************************************************************
def standardHillClimbingLocalSearch(original_lst, random_solution, size, max_iteration):
	random_solution = randomSol2(size)

	prior_residue = karp2(original_lst, random_solution, size)

	for i in range(max_iteration):
		new_random_solution = randomMove2(random_solution, size)
		

		new_residue = karp2(original_lst, new_random_solution, size)


		if new_residue < prior_residue:
			prior_residue = new_residue
			random_solution = new_random_solution

	return random_solution


def standardSimulatedAnnealingLocalSearch(original_lst, random_solution, size, max_iteration):
	# random_solution = randomSol2(size)
	random_solution2 = random_solution

	prior_residue = karp2(original_lst, random_solution, size)

	original_residue = prior_residue

	for i in range(max_iteration):
		new_random_solution = randomMove2(random_solution, size)

		new_residue = karp2(original_lst, new_random_solution, size)

		if new_residue < prior_residue:
			prior_residue = new_residue
			random_solution = new_random_solution
		else:
			if random.random() < exp(-(new_residue - prior_residue)/float(t_iteration(i))):
				prior_residue = new_residue
				random_solution = new_random_solution

		if prior_residue < original_residue:
			random_solution2 = random_solution
			original_residue = prior_residue

	return random_solution2

def standardRepeatedRandomLocalSearch(original_lst, random_solution, size, max_iteration):
	# random_solution = randomSol2(size)

	prior_residue = karp2(original_lst, random_solution, size)

	for i in range(max_iteration):
		new_random_solution = randomSol2(size)

		new_residue = karp2(original_lst, new_random_solution, size)

		

		if new_residue < prior_residue:
			prior_residue = new_residue
			random_solution = new_random_solution


	return random_solution

#****************************************************************************************

random.seed()
size = 100
max_iteration = 25000

writeFile("input.txt", size)
original_lst = readFile("input.txt", size)

random_solution_prepartioning = randomSol(size)
random_solution1 = copy.deepcopy(random_solution_prepartioning)
random_solution2 = copy.deepcopy(random_solution_prepartioning)
random_solution3 = copy.deepcopy(random_solution_prepartioning) 

print("prepartitioned repeated random results")
solution1 = repeatedRandomLocalSearch(original_lst, random_solution1, size, max_iteration)
# print solution1
# print newInput(original_lst, solution1, size)
print(karp(newInput(original_lst, solution1, size), size))

print("prepartitioned hill climbing results")
solution2 = hillClimbingLocalSearch(original_lst, random_solution2, size, max_iteration)
# print solution2
# print newInput(original_lst, solution2, size)
print(karp(newInput(original_lst, solution2, size), size))

print("prepartitioned simulated annealing results")
solution3 = simulatedAnnealingLocalSearch(original_lst, random_solution3, size, max_iteration)
# print solution3
# print newInput(original_lst, solution3, size)
print(karp(newInput(original_lst, solution3, size), size))


random_solution_standarized = randomSol2(size)
random_solution1 = copy.deepcopy(random_solution_standarized)
random_solution2 = copy.deepcopy(random_solution_standarized)
random_solution3 = copy.deepcopy(random_solution_standarized)


print("standard repeated random results")
standard_solution1 = standardRepeatedRandomLocalSearch(original_lst, random_solution1, size, max_iteration)
# print standard_solution1
print(karp2(original_lst, standard_solution1, size))

print("standard hill climbing results")
standard_solution2 = standardHillClimbingLocalSearch(original_lst, random_solution2, size, max_iteration)
# print standard_solution2
print(karp2(original_lst, standard_solution2, size))

print("standard simulated annealing results")
standard_solution3 = standardSimulatedAnnealingLocalSearch(original_lst, random_solution3, size, max_iteration)
# print standard_solution3
print(karp2(original_lst, standard_solution3, size))







